pkg/backup/backup.go

```diff
@@ -179,94 +180,128 @@ func (b *Backup) Run() {
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
+func GetSnap(ctx context.Context, kclient kubernetes.Interface, tls *tls.Config, namespace, clusterName string) (rc io.ReadCloser, etcdVersion string, rev int64, err error) {
+	pods, err := getRunningPods(kclient, namespace, clusterName)
+	if err != nil {
+		return nil, "", 0, fmt.Errorf("no running etcd pods found")
+	}
+
+	m, rev := getMemberWithMaxRev(pods, tls)
+	if m == nil {
+		return nil, "", 0, fmt.Errorf("no reachable member")
+	}
 
 	cfg := clientv3.Config{
 		Endpoints:   []string{m.ClientURL()},
 		DialTimeout: constants.DefaultDialTimeout,
-		TLS:         b.etcdTLSConfig,
+		TLS:         tls,
 	}
 	etcdcli, err := clientv3.New(cfg)
 	if err != nil {
-		return fmt.Errorf("failed to create etcd client (%v)", err)
+		return nil, "", 0, fmt.Errorf("failed to create etcd client (%v)", err)
 	}
 	defer etcdcli.Close()
 
-	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
 	resp, err := etcdcli.Maintenance.Status(ctx, m.ClientURL())
-	cancel()
 	if err != nil {
-		return err
+		return nil, "", 0, fmt.Errorf("failed to get member status (%v)", err)
 	}
 
-	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
-	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+	ctx, cancel := context.WithTimeout(ctx, constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err = etcdcli.Maintenance.Snapshot(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to receive snapshot (%v)", err)
+		return nil, "", 0, fmt.Errorf("failed to receive snapshot (%v)", err)
 	}
-	defer cancel()
-	defer rc.Close()
+	return rc, resp.Version, rev, nil
+}
 
-	n, err := b.be.save(resp.Version, rev, rc)
+func (b *Backup) getLatestBackupRev() int64 {
+	// If there is any error, we just exit backup sidecar because we can't serve the backup any way.
+	name, err := b.be.getLatest()
 	if err != nil {
-		return err
+		logrus.Fatal(err)
 	}
-
-	bs := backupapi.BackupStatus{
-		CreationTime:     time.Now().Format(time.RFC3339),
-		Size:             toMB(n),
-		Version:          resp.Version,
-		Revision:         rev,
-		TimeTookInSecond: int(time.Since(start).Seconds() + 1),
+	if len(name) == 0 {
+		return 0
 	}
-	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
-	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
-		b.recentBackupsStatus = b.recentBackupsStatus[1:]
+	rev, err := getRev(name)
+	if err != nil {
+		logrus.Fatal(err)
 	}
+	return rev
+}
+
+func (b *Backup) getLatestBackupStatus() backupapi.BackupStatus {
+	return b.recentBackupsStatus[len(b.recentBackupsStatus)-1]
+}
 
-	return nil
+func getRunningPods(kclient kubernetes.Interface, namespace, clusterName string) ([]*v1.Pod, error) {
+	podList, err := kclient.Core().Pods(namespace).List(k8sutil.ClusterListOpt(clusterName))
+	if err != nil {
+		return nil, err
```

Comment:
Wrap the error and then use it above in GetSnap():
`return nil, fmt.Errorf("failed to list pods: %v", err)`